### PR TITLE
Rust SDK

### PR DESCRIPTION
**Pull request by ErisMik**
- Adds a Rust wrapper for the Picovoice SDK as well as both file and mic demos

- There are some unit tests for the wrapper. Manual testing was done on a a Dell XPS laptop for Linux

- You can verify the changes by installing Rust (https://rustup.rs/) and Clang (`libclang-dev` linux / https://github.com/llvm/llvm-project/releases/tag/llvmorg-12.0.1 windows), and running `cargo run --release -- <args>` in either of the demo dirs.

### CODE REVIEW

File: demo/rust/README.md [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+## Installation
+
+MicDemo uses [miniaudio-rs](https://github.com/ExPixel/miniaudio-rs) for cross-platform audio capture. It uses `bindgen` and therefore requires `clang` to be installed and on the path. Use the [`Bindgen` docs](https://rust-lang.github.io/rust-bindgen/requirements.html) for instructions on how to install `clang` for various Operating Systems and distros.
+
+## Usage
+
+NOTE: The working directory for the following `Cargo` commands is:
+
+```console
+rhino/demo/rust/filedemo  # File Demo
```
Comment (laves) [inline demo/rust/README.md:14]: *picovoice/demo

File: demo/rust/README.md [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+rhino/demo/rust/micdemo  # Mic Demo
```
Comment (laves) [inline demo/rust/README.md:15]: *picovoice/demo

File: demo/rust/README.md [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+```
+
+To see all available arguments, use the `--help` flag:
+```console
+cargo run --release -- --help
+```
+
+### Microphone Demo
+
+It opens an audio stream from a microphone and detects utterances of a give wake word(s). The following processes
```
Comment (laves) [inline demo/rust/README.md:38]: *The microphone demo

File: demo/rust/filedemo/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+    }
+}
+
+fn main() {
+    let matches = App::new("Picovoice Rhino Rust File Demo")
+        .arg(
+            Arg::with_name("input_audio_path")
+            .long("input_audio_path")
+            .value_name("PATH")
+            .help("Path to input audio file (mono, WAV, 16-bit, 16kHz)")
```
Comment (laves) [inline demo/rust/filedemo/src/main.rs:95]: do we enforce the audio format requirements in the demo? I don't know much about the audio library you're using, but if there's a way to let the user know their file isn't valid, that would be ideal

File: sdk/rust/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,25 @@
+[package]
+name = "picovoice"
+version = "0.0.1"
+edition = "2018"
+description = "The Rust version of Picovoice's SDK"
```
Comment (laves) [inline sdk/rust/Cargo.toml:5]: maybe "Rust SDK for Picovoice's voice recognition platform" would sell better

File: sdk/rust/README.md [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    context_path,
+    inference_callback,
+).init().expect("Failed to create picovoice");
+```
+Upon detection of wake word defined by `keyword_path` it starts inferring user's intent from the follow-on voice command within
+the context defined by the file located at `context_path`. `keyword_path` is the absolute path to
+[Porcupine wake word engine](https://github.com/Picovoice/porcupine) keyword file (with `.ppn` suffix).
+`contextPath` is the absolute path to [Rhino Speech-to-Intent engine](https://github.com/Picovoice/rhino) context file
+(with `.rhn` suffix). `wake_word_callback` is invoked upon the detection of wake phrase and `inference_callback` is
+invoked upon completion of follow-on voice command inference.
```
Comment (laves) [inline sdk/rust/README.md:62]: Can you also add a section here about overwriting default models and sensitivities?

**Review by laves — CHANGES_REQUESTED** [review]

File: demo/rust/filedemo/src/main.rs (hunk 4 above)
Comment (ErisMik) [reply demo/rust/filedemo/src/main.rs:95]: Technically the audio reading library used here can handle any standard audio file format (and any sample type wrt. WAV), the only thing we need is the 16kHz sample rate which we do check for

Comment (laves) [reply demo/rust/filedemo/src/main.rs:95]: but no matter what they input, we get our required format for sure? like if they input a 16kHz mp3 do we report an error or take it?

Comment (ErisMik) [reply demo/rust/filedemo/src/main.rs:95]: We will take it and it will work!

Comment (laves) [reply demo/rust/filedemo/src/main.rs:95]: nifty! In that case, can you either update the help to indicate that more formats are accepted, or block them from sending anything but WAV?

Comment (ErisMik) [reply demo/rust/filedemo/src/main.rs:95]: Actually I take it back, I played around a bit more with the formats just to double check and it seems I just got lucky with the ones I tested before, not everything "just works". I'll just restrict it to what's in the help as suggested.

Comment (ErisMik) [reply demo/rust/filedemo/src/main.rs:115]: I'll make this same change in the Rhino & Porcupine demos also

**Review by laves — APPROVED** [review]

**Pull request merged**